File: sdks/python/engram_client/resources/spatial.py
"""Spatial navigation resource mixin (Method of Loci)."""

from __future__ import annotations

from typing import Any

from .base import ResourceMixin


class SpatialMixin(ResourceMixin):
    """Mnemonic spatial memory operations (Palace, Wings, Rooms, Drawers)."""

    async def palace_navigate(
        self,
        *,
        workspace: str = "default",
        wing: str | None = None,
    ) -> dict[str, Any]:
        """Navigate the Memory Palace to discover wings, rooms, and drawer counts."""
        params: dict[str, Any] = {"workspace": workspace}
        if wing is not None:
            params["wing"] = wing
        return await self._mcp_call("palace_navigate", params)

    async def room_search(
        self,
        wing: str,
        query: str,
        *,
        room: str | None = None,
        limit: int = 10,
        workspace: str | None = None,
    ) -> dict[str, Any]:
        """Search memories scoped within a specific wing and room."""
        params: dict[str, Any] = {
            "wing": wing,
            "query": query,
            "limit": limit,
        }
        if room is not None:
            params["room"] = room
        if workspace is not None:
            params["workspace"] = workspace
        return await self._mcp_call("room_search", params)

    async def drawer_open(self, memory_id: int) -> dict[str, Any]:
        """Open a memory drawer by ID to inspect full verbatim content and metadata."""
        return await self._mcp_call("drawer_open", {"id": memory_id})

    async def palace_visualize(
        self,
        *,
        workspace: str = "default",
        wing: str | None = None,
        format_: str = "html",
        output_path: str | None = None,
    ) -> dict[str, Any]:
        """Generate or export a memory palace topological visualization in HTML, ASCII, SVG, Mermaid, or JSON format."""
        params: dict[str, Any] = {
            "workspace": workspace,
            "format": format_,
        }
        if wing is not None:
            params["wing"] = wing
        if output_path is not None:
            params["output_path"] = output_path
        return await self._mcp_call("palace_visualize", params)

File: src/mcp/tools/catalog/spatial.rs
//! MCP tool definitions: Spatial Navigation ("Method of Loci").
//! Mnemonic spatial abstractions (Palace, Wings, Rooms, Drawers) over Engram workspaces and scopes.

use crate::mcp::protocol::ToolAnnotations;
use crate::mcp::tools::{ToolDef, ToolTier};

#[allow(dead_code)]
pub const TOOLS: &[ToolDef] = &[
    ToolDef {
        name: "palace_navigate",
        description: "Navigate the Memory Palace: inspect active wings (top-level domains), rooms (sub-topics), and memory counts for spatial orientation.",
        schema: r#"{
            "type": "object",
            "properties": {
                "workspace": {
                    "type": "string",
                    "description": "Target workspace (Palace). Defaults to 'default'."
                },
                "wing": {
                    "type": "string",
                    "description": "Optional wing filter to inspect rooms inside a specific wing."
                }
            }
        }"#,
        annotations: ToolAnnotations::read_only(),
        tier: ToolTier::Standard,
    },
    ToolDef {
        name: "room_search",
        description: "Search memories scoped within a specific spatial room and wing using hybrid retrieval.",
        schema: r#"{
            "type": "object",
            "properties": {
                "wing": {
                    "type": "string",
                    "description": "Target wing (domain/project)."
                },
                "room": {
                    "type": "string",
                    "description": "Optional room (subtopic/component)."
                },
                "query": {
                    "type": "string",
                    "description": "Search query."
                },
                "limit": {
                    "type": "integer",
                    "default": 10,
                    "description": "Maximum number of drawer memories to return."
                },
                "workspace": {
                    "type": "string",
                    "description": "Optional workspace name."
                }
            },
            "required": ["wing", "query"]
        }"#,
        annotations: ToolAnnotations::read_only(),
        tier: ToolTier::Standard,
    },
    ToolDef {
        name: "drawer_open",
        description: "Open a specific memory drawer by ID to read its full verbatim content and metadata.",
        schema: r#"{
            "type": "object",
            "properties": {
                "id": {
                    "type": "integer",
                    "description": "Drawer (memory) ID to open."
                }
            },
            "required": ["id"]
        }"#,
        annotations: ToolAnnotations::read_only(),
        tier: ToolTier::Standard,
    },
    ToolDef {
        name: "palace_visualize",
        description: "Generate or export a topological visualization of the memory palace in interactive HTML, ASCII floorplan, vector SVG, Mermaid mindmap, or Cytoscape JSON.",
        schema: r#"{
            "type": "object",
            "properties": {
                "workspace": {
                    "type": "string",
                    "description": "Target workspace (Palace). Defaults to 'default'."
                },
                "wing": {
                    "type": "string",
                    "description": "Optional wing filter to isolate a specific palace wing."
                },
                "format": {
                    "type": "string",
                    "enum": ["html", "ascii", "svg", "mermaid", "json"],
                    "default": "html",
                    "description": "Visualization export format ('html', 'ascii', 'svg', 'mermaid', 'json'). Defaults to 'html'."
                },
                "output_path": {
                    "type": "string",
                    "description": "Optional local file path to save the rendered output."
                }
            }
        }"#,
        annotations: ToolAnnotations::read_only(),
        tier: ToolTier::Standard,
    },
];
